### DISCUSSION TITLE
Packets ID Finding

### DISCUSSION THREAD

**Discussion by KagefumiMerry** [PhantasyServer/pso2-protocol-lib#1] [General]
Market related packets (ID 0x2D)
Market Item List packets
(0x2D, 0x03)  Maybe each item for 132 bytes (NGS) or 128 bytes (classic one)
0x0 owner player ID?
0x1C item main category id
0x1E item sub category id
0x22 item id
(These could be referee on https://github.com/CRC32EX/PSO2Data/blob/master/items.csv )
0x74 No. of items
0x78 selling price

Market Item shop owner?
(0x2D, 0x0C)

**Comment by AntonnMal** [comment]
Added those packets (50c3266). Also it seems that 0x34 is also related to the market. Maybe only for top sellers, etc, I'm not sure on that.

**Comment by KagefumiMerry** [comment]
Additional to Login packets [0x11] list:
[0x11 0xE1] seems is the packets contain the ICE files while login. 

the LoadItemAttributes [Id(0x0F, 0x00)] packets when login, could be combined to single large ICE file (item_parameter), however i am still not sure how to process the file to id list 
https://raw.githubusercontent.com/CRC32EX/PSO2Data/master/items.csv

(I just see that Lapig : "i think thats probably the same data as whats in the 11e1 packets the other ice data u get wen u login" )

Did the packet format like the LoadItemPacket [Id(0x0F, 0x30)] so we could get the ItemId directly and map to item name?

For Player Spawn packet [Id(0x08, 0x04)]:
packetdata[4] == 0x4 (thus id like 0x04, 0x08, 0x04, 0x00 ) indicate the player is in NGS block 
the size of Spawn packet in NGS is not fixed, so need to do pattern check or other ways to identify the information

Staring from 0x60 with each 0x30 (48bytes) block to check pattern with the same start (0x0-0x2) and end with 0x4 (at 0x25)
the "start" of each block should be like 7000 (00 00 58 1B), 5000 (00 00 88 13), 4000(00 00 A0 00)

After 0x14 bytes after these repeated blocks, arksCharacterName and 
arksCharacterName + 0x20 -> arkslook (768bytes)
arkslook + 0x610 ( or 0x60C before 20231011 update) ->arksUserName

Here is my current findings.

Not sure it is useful or not but since there is no good place to put these information (even in PSO2mod-discord).

**Reply by AntonnMal** [reply]
> the LoadItemAttributes [Id(0x0F, 0x00)] packets when login, could be combined to single large ICE file (item_parameter), however i am still not sure how to process the file to id list

I have _some_ information about how to parse it in classic ([structure](https://github.com/PhantasyServer/pso2-protocol-lib/blob/master/src/protocol/models/item_attrs.rs)). NGS as far as i know is similar, it just has some extra fields in some objects.

> For Player Spawn packet [Id(0x08, 0x04)]:
packetdata[4] == 0x4 (thus id like 0x04, 0x08, 0x04, 0x00 ) indicate the player is in NGS block

This is very annoying (it's not hard to implement, but I'd rather not). Sometime later I think I'll look into that.

> Not sure it is useful or not but since there is no good place to put these information (even in PSO2mod-discord).

That's fine, this repo is also for documenting the protocol, feel free to post your finding here.

**Comment by KagefumiMerry** [comment]
> > the LoadItemAttributes [Id(0x0F, 0x00)] packets when login, could be combined to single large ICE file (item_parameter), however i am still not sure how to process the file to id list
> 
> I have _some_ information about how to parse it in classic ([structure](https://github.com/PhantasyServer/pso2-protocol-lib/blob/master/src/protocol/models/item_attrs.rs)). NGS as far as i know is similar, it just has some extra fields in some objects.

I try to get updated with current NGS item_attrs, but since I have no too much Rust background to understand the Procedural Macros , thus I try it by hand in Hex-editor.
 
In 202403 JP updates, NGS version item_parameter.bin maybe like this (guessing but not verify every entry, maybe completely wrong):
the number of the items seems changed to uint32

The classic seems have another item_parameter.bin while login, and also I need a way to get the items name from packets (did the id 0x31 should be the name list? as [Id(0x31, 0x05)] contain the achievement titles )..

Also id[0x40, 0x2A] should be the friends related list (online list?)...

| No | Items | bytes_of_each_item | start_pos |                               |
|----|-------|-------|-----------|-------------------------------|
| 1  | 4691  | 60    | 20        | WeaponAttrs                   |
| 2  | 5226  | 36    | 281484    | HumanCostume                  |
| 3  | 1001  | 32    | 469624    | CastPart                      |
| 4  | 55624 | 40    | 501660    | Consumable? [Ba] etc.         |
| 5  | 533   | 96    | 2726624   |                               |
| 6  | 765   | 72    | 2777796   | units                         |
| 7  | 1238  | 84    | 2832880   | camouflage                    |
| 8  | 3     | 20    | 2936876   | Time-Limited Enhancement Data?? |
| 9  | 123   | 64    | 2936940   | Eggs                          |
| 10 | 474   | 52    | 2944816   |                               |
| 11 | 8     | 20    | 2969468   | Bonus Key                     |
| 12 | 236   | 52    | 2969632   |                               |
| 13 | 52    | 112   | 2981908   |                               |
| 14 | 106   | 128   | 2987736   | Skill Ring                    |
| 15 | 357   | 20    | 3001308   |                               |
| 16 | 109   | 20    | 3008452   |                               |
| 17 | 1     | ???   | 3010636   |  From this block, i could not understand the information there                             |
| 19 | 2909  | 124   | 3014334   |                               |
| 20 | 3     | 44    | 3375050   |                               |

**Reply by AntonnMal** [reply]
> ...also I need a way to get the items name from packets...

At least in classic item names are loaded when needed using 0x0F, 0x30 packet. So if the server sends a particular item id from any packet, it also sends this [LoadItem](https://github.com/PhantasyServer/pso2-protocol-lib/blob/fc8568936078704469c2395a0e2dc35e383b09c3/src/protocol/items.rs#L537) packet if the client sees this item for the first time since logging on.

**Comment by KagefumiMerry** [comment]
> > ...also I need a way to get the items name from packets...
> 
> At least in classic item names are loaded when needed using 0x0F, 0x30 packet. So if the server sends a particular item id from any packet, it also sends this [LoadItem](https://github.com/PhantasyServer/pso2-protocol-lib/blob/fc8568936078704469c2395a0e2dc35e383b09c3/src/protocol/items.rs#L537) packet if the client sees this item for the first time since logging on.

0x0F, 0x30 packet seems last hope on item names (but seems only game available/player owned items could get 0x0F, 0x30 packet)

I am also looking at 9 0x11, 0xe1 packets (ICE packs, a bit away from packet-lib usages) 

But I need time to see how the UTF16 string store in ICE files (seem not like .text / .cmx file format?)
